dart/lib/src/protocol/access_aware_map.dart

```diff
@@ -0,0 +1,50 @@
+import 'dart:collection';
+
+import 'package:meta/meta.dart';
+
+@internal
+class AccessAwareMap<String, V> extends MapBase<String, V> {
+  AccessAwareMap(this._map);
+
+  final Map<String, V> _map;
+  final Set<String> _accessedKeys = {};
+
+  Set<String> get accessedKeys => _accessedKeys;
+
+  @override
+  V? operator [](Object? key) {
+    if (key is String) {
+      _accessedKeys.add(key);
+    }
+    return _map[key];
+  }
+
+  @override
+  void operator []=(String key, V value) {
+    _map[key] = value;
+  }
+
+  @override
+  void clear() {
+    _map.clear();
+    _accessedKeys.clear();
+  }
+
+  @override
+  Iterable<String> get keys => _map.keys;
+
+  @override
+  V? remove(Object? key) {
+    return _map.remove(key);
+  }
+
+  Map<String, dynamic>? notAccessed() {
```

Comment:
maybe we could improve the naming of the function?

`unknownFields`? or something like that

Reply:
The naming is in line with the `AccessAwareMap` class name, so it fits in this context. We just use it for determining unknwon fields.